### DISCUSSION TITLE
Widget Support

### DISCUSSION THREAD

**Discussion by ajmeese7** [DustinBrett/daedalOS#48] [Feature Requests]
Hey Dustin, I was just wondering if you have considered adding widget support to your site. I think that would be an awesome addition!

**Comment by DustinBrett** [comment]
Great idea! I don't think it's been mentioned or written down yet so I will add it. Any specific widgets you'd like to see? I was thinking when I made my [virtual assistant](https://www.smore.com/clippy-js) I would make a widget like interface because he will kinda float.

**Comment by ajmeese7** [comment]
I think that's a great idea, I've been planning to add one of those to my wife's website! One cool idea would definitely be a visitor count widget, I have a simple PHP script that I use for one on one project I'm currently working on. It's an application, so not quite a widget, but the image of how mine is set up is attached below:

![image](https://user-images.githubusercontent.com/17814535/150621745-6f9be845-8559-4cc0-9030-d9722967e1ad.png)

I think a minimal design on yours with just pretty text would be nice, and you could add other simple stuff to start out like a clock and potentially an image slideshow. I'd be happy to send the code if you're interested, but I'm sure you can make something better.

I love your site, I look forwards to continuing to follow your work on it and seeing how it grows!

**Comment by DustinBrett** [comment]
Thanks for the ideas and mockup! I will keep this issue updated as I make progress toward Widget support.

**Comment by triplu-zero** [comment]
Here is my first attempt at this issue in JavaScript:
 - transcendental objects
 
I found this abstraction useful because at the time in JS it was a good way to acces more OS features in an elegant way. And since you have an OS maybe you can resurrect this ideea and improve upon it.
Note that the TObject() function is just a mockup in this old implementation: I needed threads, a shell with processes, and OS widgets and at one moment in time it worked as intended.

Please feel free to incorporate all the ideas that you find useful. I suggest that you use this index.html file inside your OS as a first example of a working widget once you get familiar with it and squeeze the bugs you find, and modify it to suit your needs:
 - it uses the TEXTAREA element which has limits, and you already included a full fledged editor which could be inserted, with the proper abstractions, and replace it as a 'transcendental object'
 - it probably can use python scripts included as TObjects
 - spawn processes and return a TObject through which they can be controlled

https://github.com/triplu-zero/triplu-zero.github.io/blob/main/index.html

The REPL is here (Type Ctrl+~ to toggle input modes. Note that the buttons have old code that must be fixed.):
https://triplu-zero.github.io/

![Transcendental_Object_0001](https://user-images.githubusercontent.com/84599818/151027632-5e4c36ef-d9da-4862-8c0c-90911658592b.png)

**Comment by DustinBrett** [comment]
Thanks for this. My current thinking is still that widgets can be similar to "apps" in that it's mostly going to be giving developers something where you give it jsx and you say if you want it in a window or floating, and if it's floating we could call it a widget and perhaps have other details to denote that it's slightly different than a window.

**Reply by ajmeese7** [reply]
That's a great way of handling it, I like that!
